src/Test/StrictCheck/Observe.hs

```diff
@@ -173,34 +183,42 @@ entangle a = do
 -- [1,2]
 -- >>> d
 -- "1 : 2 : _ : _"
+
+-- NOTE: There are a two properties we care about:
+--
+--   1. Running the Demand action should always result in a consistent state
+--      and not depend on when its result is forced.
+--   2. We want to evaluate the input as little as possible. Importantly
+--      running the demand action should never force any previously unforced
+--      parts of the input. This is especially important for the observe
+--      functions and the hardest thing to get right.
 entangleShape :: Shaped a => a -> IO (a, IO (Demand a))
-entangleShape = entangleShape' . project I
-  -- NOTE: There are a two properties we care about:
-  --
-  --   1. Running the Demand action should always result in a consistent state
-  --      and not depend on when its result is forced.
-  --   2. We want to evaluate the input as little as possible. Importantly
-  --      running the demand action should never force any previously unforced
-  --      parts of the input. This is especially important for the observe
-  --      functions and the hardest thing to get right.
+entangleShape ((project I -> s) :: a) =
+   fmap (\(~(~(a, _), da)) -> (a, Wrap <$> (traverse snd =<< da)))
+   . entangle =<<
+   -- We need to use unsafeInterleaveIO here so that we do not force the value
+   -- by matching on it when we bind the result above to entangle it.
+   --
+   -- Using unsafeInterleaveIO here is safe, as the result of the IO action does
+   -- not depend on when it is performed and it doesn't matter if it is never
+   -- performed, if it's value is not forced.
+   unsafeInterleaveIO entangledChildren
   where
-    entangleShape' :: forall a. Shaped a => Shape a I -> IO (a, IO (Demand a))
-    entangleShape' s =
-      (fmap (bimap fst (fmap Wrap . traverse snd =<<)) . entangle =<<)
-      . unsafeInterleaveIO $
-        match @a s s (\flatA _ ->
-          fmap (\flat' ->
-            let a = embed unI
-                  . unflatten
-                  . mapFlattened @Shaped (I . demanded)
-                  $ flat'
-                ds = fmap unflatten
-                   . traverseFlattened @Shaped getDemand
-                   $ flat'
-            in (a, ds))
-          . traverseFlattened @Shaped
-            ((uncurry WithDemand <$>) . entangleShape . unI)
-          $ flatA)
+    -- The to be entangled value with all its recursive children entangled. We
+    -- still need to entangle the value itself.
+    entangledChildren :: IO (a, IO (Shape a Demand))
```

Comment:
Instead of using `match` directly, I think there's a way to express this using `translateA`, which is the `traverse` equivalent for `Shape`s. I'll check this out later.

Reply:
Good idea. Implemented it on my branch.